### ISSUE TITLE
All traffic inbound to the cluster fails after upgrading from 1.15 to 1.16 (GKE-only)

### ISSUE THREAD

**Issue by itsthatriver** [cilium/cilium#35977]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.16.0 and lower than v1.17.0

### What happened?

After upgrading Cilium from 1.15 to 1.16 on GKE, all externally exposed services became inaccessible. Traffic inside the cluster was not impacted, and Cilium status was all healthy. This included Primitive LoadBalancer Services, Ingresses, and Gateway API.

### How can we reproduce the issue?

1. Install Cilium with Helm on GKE (we're on 1.30). This should be on a Legacy Datapath cluster (not Dataplane V2)
2. Values:
```
agentNotReadyTaintKey: ignore-taint.cluster-autoscaler.kubernetes.io/cilium-agent-not-ready
aksbyocni:
  enabled: false
authentication:
  mutual:
    spire:
      enabled: true
      install:
        enabled: true
        existingNamespace: true
        namespace: kube-system
bpf:
  masquerade: true
cni:
  binPath: /home/kubernetes/bin
devices: eth+
encryption:
  enabled: true
  nodeEncryption: true
  type: wireguard
envoy:
  enabled: true
gatewayAPI:
  enableAlpn: true
  enableAppProtocol: true
  enabled: true
  secretsNamespace:
    create: false
    name: kube-system
hubble:
  listenAddress: :4244
  metrics:
    enabled:
    - dns
    - drop
    - tcp
    - flow
    - icmp
    - http
  relay:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: pool
              operator: In
              values:
              - default
              - control
      podAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
            matchLabels:
              k8s-app: cilium
          topologyKey: kubernetes.io/hostname
    enabled: true
  tls:
    auto:
      method: cronJob
  ui:
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: pool
              operator: In
              values:
              - default
              - control
    enabled: true
ingressController:
  enabled: false
ipam:
  mode: kubernetes
k8sServiceHost: REDACTED
k8sServicePort: 443
kubeProxyReplacement: true
l7Proxy: true
loadBalancer:
  serviceTopology: true
localRedirectPolicy: true
nodeinit:
  enabled: true
  reconfigureKubelet: true
  removeCbrBridge: true
operator:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: pool
            operator: In
            values:
            - default
            - control
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchLabels:
            io.cilium/app: operator
        topologyKey: kubernetes.io/hostname
  prometheus:
    enabled: true
prometheus:
  enabled: true
upgradeCompatibility: "1.9"
wellKnownIdentities:
  enabled: true
```
3. Create an nginx deployment and a LoadBalancer in front of it, attempt to curl the LoadBalancer IP.

### Cilium Version

This was reproduced with 1.16.1 and 1.16.3. Once I downgraded to 1.15.9 the problem immediately went away.

### Kernel Version

Linux 6.1.100+ #1 SMP PREEMPT_DYNAMIC Sat Aug 24 16:19:44 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

v1.30.5-gke.1014003

This was confirmed on two separate GKE clusters with different external services.

### Regression

This is a regression. The exact same config worked on 1.15.9

### Sysdump

I'll grab this later, I had to fix the cluster because of a deadline.

### Relevant log output

```shell
I was not able to find any logs that appeared relevant.
```

### Anything else?

I did find that the external service ports were open (tested with `nc -zv`), but curl resulted in `refused to connect`.

### Cilium Users Document

- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by jrife** [comment]
@thejosephstevens Do you have more detail on the exact `gcloud` command you ran to set up your cluster(s) as well as the `helm` command you used to install Cilium? At a glance, your values file seems to be missing some configuration mentioned [here](https://docs.cilium.io/en/stable/installation/k8s-install-helm/) under the GKE tab.

**Comment by thejosephstevens** [comment]
So we intentionally don't run cilium on GKE in native routing mode, as we ran into issues both with race conditions on identity creation (where an identity would be created and a packet would arrive prior to the identity being synced everywhere - this issue doesn't exist in tunneling because the identity details are carried in the packet in vxlan tunneling mode), and historically we have needed to integrate with legacy networks, and native routing drove our IP footprint through the roof (256 IPs per node vs 1). We've been running with this basic configuration (plus features getting added over time) for about 5 years since Cilium 1.9. I didn't see anything other than the nativeRouting config, and the IPV4NativeRoutingCIDR, was there something else there that you were expecting?

For our cluster config, I'd have to work my way backward to a gcloud command (we terraform everything), but here's the Terraform config:
```
locals {
  cilium_taints = [{
    key    = "ignore-taint.cluster-autoscaler.kubernetes.io/cilium-agent-not-ready"
    value  = "true"
    effect = "NO_EXECUTE"
  }]
}

resource "google_container_cluster" "main" {
  project    = var.project
  name       = "cluster-main"
  network    = local.qualified_network_name
  subnetwork = local.subnet.name
  depends_on = [google_compute_router_nat.nat] # provisioned elsewhere

  release_channel {
    channel = "UNSPECIFIED"
  }
  min_master_version = "1.30"
  location = "us-central-1"

  # Not using the default node pool as it's less configurable than an attached
  # one (where we can use pre-emptible instances among other optimizations).
  remove_default_node_pool = true
  initial_node_count       = "1"
  node_config {}

  # Workload identity is specifically disabled because it deploys netd and was fighting with cilium

  cluster_autoscaling {
    enabled             = false
    autoscaling_profile = "OPTIMIZE_UTILIZATION"
  }

  logging_service    = "none"
  monitoring_service = "none"

  datapath_provider = "LEGACY_DATAPATH"
  network_policy {
    enabled  = false
    provider = null
  }
  addons_config {
    http_load_balancing {
      disabled = true
    }

    network_policy_config {
      disabled = true # this is handled by cilium rather than gke
    }

    dns_cache_config {
      enabled = false # dns cache handling is custom to be compatible with cilium
    }

    gce_persistent_disk_csi_driver_config {
      enabled = true
    }
  }

  master_auth {
    client_certificate_config {
      issue_client_certificate = true
    }
  }

  private_cluster_config {
    enable_private_endpoint = "false"
    enable_private_nodes    = "true"
    master_ipv4_cidr_block  = 192.168.0.0/28
  }

  master_authorized_networks_config {
    cidr_blocks {
      cidr_block   = "0.0.0.0/0"
      display_name = "World"
    }
  }

  ip_allocation_policy {
    cluster_ipv4_cidr_block       = 10.0.0.0/14
    services_ipv4_cidr_block      = 192.168.2.0/23
  }

  monitoring_config {
    managed_prometheus {
      enabled = false
    }
  }
}

resource "google_container_node_pool" "default_pool" {
  project        = var.project
  name           = "cluster-default-pool"
  location       = "us-central-1"
  node_locations = "us-central-1f"
  cluster        = google_container_cluster.main.name
  depends_on     = [google_compute_router_nat.nat, google_service_account.gke_nodes]

  version = "1.30"

  node_config {
    spot            = false
    dynamic "taint" {
      for_each = local.cilium_taints
      content {
        key    = taint.value["key"]
        value  = taint.value["value"]
        effect = taint.value["effect"]
      }
    }
    workload_metadata_config {
      mode = "GCE_METADATA"
    }
    image_type   = "COS_CONTAINERD"
  }
  initial_node_count = 1
  autoscaling {
    min_node_count = 0
    max_node_count = 3
  }
  management {
    auto_repair  = true
    auto_upgrade = true
  }
}
```

For the sysdump, I've got that below.
I made a bunch of requests from `50.226.254.10`, which were aimed at a service with a public IP of `35.222.230.213`, an internal service cluster IP of `192.168.2.40`, and target pod IPs of `10.0.6.25` and `10.0.9.118`, and a target security identity of `4863`.

[cilium-sysdump-20241114-230825.zip](https://github.com/user-attachments/files/17770414/cilium-sysdump-20241114-230825.zip)

Let me know if you need anything else!

**Comment by squeed** [comment]
I see the traffic entering the node and being correctly translated to the pod IP:

```
{"flow":{"time":"2024-11-15T06:08:00.085659011Z","uuid":"41ce555f-4e5e-454b-be3f-bbc899e92062","verdict":"FORWARDED","ethernet":{"source":"92:89:bb:d8:3d:18","destination":"96:c3:af:34:4c:89"},"IP":{"source":"50.226.254.10","destination":"10.0.9.118","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":50312,"destination_port":8443,"flags":{"SYN":true}}},"source":{"identity":16777233,"labels":["cidr:0.0.0.0/1","reserved:world"]},"destination":{"ID":1281,"identity":4863,"namespace":"default","labels":["k8s:app=nginx-lb","k8s:ascend.io/cell=dedicated","k8s:ascend.io/external-service=true","k8s:ascend.io/git-sha=65c1f38","k8s:ascend.io/productTier=dedicated","k8s:io.cilium.k8s.namespace.labels.cell=default","k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=default","k8s:io.cilium.k8s.namespace.labels.trusted-traffic=true","k8s:io.cilium.k8s.policy.cluster=default","k8s:io.kubernetes.pod.namespace=default"],"pod_name":"nginx-lb-dp-6b8b55cd7f-xsccd","workloads":[{"name":"nginx-lb-dp","kind":"Deployment"}]},"Type":"L3_L4","node_name":"gke-gcp-staging-main-gcp-staging-cont-e6b8d0b7-94t5","node_labels":["beta.kubernetes.io/arch=amd64","beta.kubernetes.io/instance-type=n2d-highmem-8","beta.kubernetes.io/os=linux","cloud.google.com/gke-boot-disk=pd-standard","cloud.google.com/gke-container-runtime=containerd","cloud.google.com/gke-cpu-scaling-level=8","cloud.google.com/gke-logging-variant=DEFAULT","cloud.google.com/gke-max-pods-per-node=110","cloud.google.com/gke-memory-gb-scaling-level=65","cloud.google.com/gke-nodepool=gcp-staging-control-pool","cloud.google.com/gke-os-distribution=cos","cloud.google.com/gke-provisioning=spot","cloud.google.com/gke-spot=true","cloud.google.com/gke-stack-type=IPV4","cloud.google.com/machine-family=n2d","failure-domain.beta.kubernetes.io/region=us-central1","failure-domain.beta.kubernetes.io/zone=us-central1-c","gcp=gcp-staging-control-pool","kubernetes.io/arch=amd64","kubernetes.io/hostname=gke-gcp-staging-main-gcp-staging-cont-e6b8d0b7-94t5","kubernetes.io/os=linux","node-persistence=ephemeral","node.kubernetes.io/instance-type=n2d-highmem-8","pool=control","topology.gke.io/zone=us-central1-c","topology.kubernetes.io/region=us-central1","topology.kubernetes.io/zone=us-central1-c"],"event_type":{"type":5},"traffic_direction":"INGRESS","policy_match_type":1,"is_reply":false,"Summary":"TCP Flags: SYN","ingress_allowed_by":[{"name":"allow-ingress-to-nginx","namespace":"default","labels":["k8s:io.cilium.k8s.policy.derived-from=NetworkPolicy","k8s:io.cilium.k8s.policy.name=allow-ingress-to-nginx","k8s:io.cilium.k8s.policy.namespace=default","k8s:io.cilium.k8s.policy.uid=7c8c6adb-43ba-446e-b7b1-dff157b2ecea"],"revision":"19"},{"name":"default-allow-trusted-namespaces","namespace":"default","labels":["k8s:io.cilium.k8s.policy.derived-from=NetworkPolicy","k8s:io.cilium.k8s.policy.name=default-allow-trusted-namespaces","k8s:io.cilium.k8s.policy.namespace=default","k8s:io.cilium.k8s.policy.uid=7dc63204-be05-4f8d-bfd4-a31225690354"],"revision":"19"}]},"node_name":"gke-gcp-staging-main-gcp-staging-cont-e6b8d0b7-94t5","time":"2024-11-15T06:08:00.085659011Z"}
```

Now, GCP load balancers are special; they do not perform DNAT. Rather, there is a note agent that adds the load balancer in the `local` routing table:

```
cilium-bugtool-cilium-d6dn8-20241114-230825/cmd/cilium-dbg-statedb-dump.md
259:    {"Table":255,"LinkIndex":2,"Scope":254,"Dst":"35.222.230.213/32","Src":"","Gw":""},
```

And, correspondingly, I see two conntrack entries in the sysdump

```
❯ ack 50.226.254.10:50312 cilium-bugtool-cilium-d6dn8-20241114-230825
cilium-bugtool-cilium-d6dn8-20241114-230825/cmd/cilium-dbg-bpf-ct-list-global.md
1037:TCP OUT 50.226.254.10:50312 -> 10.0.9.118:8443 expires=86091 Packets=0 Bytes=0 RxFlagsSeen=0x16 LastRxReport=86081 TxFlagsSeen=0x1a LastTxReport=86080 Flags=0x0033 [ RxClosing TxClosing SeenNonSyn NodePort ] RevNAT=18 SourceSecurityID=2 IfIndex=0 BackendID=0 
2882:TCP SVC 50.226.254.10:50312 -> 35.222.230.213:443 expires=94081 Packets=0 Bytes=0 RxFlagsSeen=0x00 LastRxReport=0 TxFlagsSeen=0x1a LastTxReport=86080 Flags=0x0010 [ SeenNonSyn ] RevNAT=18 SourceSecurityID=0 IfIndex=0 BackendID=66 
3196:TCP IN 50.226.254.10:50312 -> 10.0.9.118:8443 expires=86091 Packets=0 Bytes=0 RxFlagsSeen=0x1e LastRxReport=86081 TxFlagsSeen=0x1e LastTxReport=86081 Flags=0x0013 [ RxClosing TxClosing SeenNonSyn ] RevNAT=0 SourceSecurityID=16777233 IfIndex=0 BackendID=0 
```

I wonder if something related to this has changed between v1.15 and v1.16.

If you have a v1.15 cluster lying around, do you think you could capture a conntrack entry of a **successful** flow?

**Comment by squeed** [comment]
My first guess is that we're not sending traffic back out with the right source IP. To the best of my knowledge, the ultimate source IP should be 35.222.230.213.

**Comment by dajrivera** [comment]
I had a similar issue when upgrading to 1.16.x. I was able to workaround by setting `bpf.hostLegacyRouting=true`. If I understand correctly, it is a bug in the BPF IPv4 code and will be fixed in 1.16 with #35861.

**Comment by thejosephstevens** [comment]
@dajrivera got it! Note, it did require a node roll after helm upgrade and apply, but once I did all of that my external services worked again! Thanks for the help! I did think this was weird because we're running this same setup on EKS with very few differences, so I was surprised that we encountered this problem on GKE. I'll note my findings on #35861 for visibility.

Thank you all for the help!
